### PR TITLE
APEXMALHAR-2434 Fix JMSTransactionableStore by using deterministic qu…

### PR DESCRIPTION
**Pull request by oliverwnk**
…eueName + messageSelector

@sanjaypujare please review

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java [hunk 1]
```diff
@@ -33,7 +33,10 @@
 
 /**
  * This transactionable store commits the messages sent within a window along with the windowId of the completed window
- * to JMS. This store ensures that the JMS output operator is capable of outputting data to JMS exactly once.
+ * to JMS. WindowIds will be sent to a subject specific meta-queue with the name of the form '{subject}.metadata'. Physical
+ * creation of this meta-queue is required by the JMS provider.
```
Comment (sanjaypujare) [inline library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java]: Not sure what you mean by "Physical creation of this meta-queue is required by the JMS provider."  Did you mean to say "It is the responsibility of the user to create the meta-data queue."

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java [hunk 2]
```diff
@@ -33,7 +33,10 @@
 
 /**
  * This transactionable store commits the messages sent within a window along with the windowId of the completed window
- * to JMS. This store ensures that the JMS output operator is capable of outputting data to JMS exactly once.
+ * to JMS. WindowIds will be sent to a subject specific meta-queue with the name of the form '{subject}.metadata'. Physical
```
Comment (sanjaypujare) [inline library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java]: Is "meta-queue" the right term or is "meta-data queue" preferable?

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java [hunk 3]
```diff
@@ -65,17 +70,18 @@ public long getCommittedWindowId(String appId, int operatorId)
...
     try {
-
       beginTransaction();
       BytesMessage message = (BytesMessage)consumer.receive();
-      logger.debug("Retrieved committed window message id {}", message.getJMSMessageID());
+      logger.debug("Retrieved committed window messageId: {}, messageAppOperatorIdProp: {}", message.getJMSMessageID(),
+          message.getStringProperty(APP_OPERATOR_ID));
       long windowId = message.readLong();
 
       message = getBase().getSession().createBytesMessage();
```
Comment (sanjaypujare) [inline library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java]: Can you refactor these 4 lines to make a new function as follows. Then you can call it from 3 places (getCommittedWindow, storeCommitedWindow and connect)

private void writeWindowId(long windowId)
{
    BytesMessage message = getBase().getSession().createBytesMessage();
    message.setStringProperty(APP_OPERATOR_ID, appId + "_" + operatorId);
    message.writeLong(windowId);
    producer.send(message);
}

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 4]
```diff
@@ -109,6 +112,33 @@ private void createOperator()
     outputOperator.setup(testOperatorContext);
   }
 
+  private JMSStringSinglePortOutputOperator createOperator2()
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Can you eliminate code duplication here? Lines 125-137 match 100-112 (with the exception of outputOperator id which in one case is local and in the other class member and store vs store2 etc). Try to have only one method "createOperator" that deals with outputOperator,store vs outputOperator2,store2 or at least eliminate duplication and remove outputOperator name confusion

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 5]
```diff
@@ -157,6 +187,35 @@ public void storeRetreiveTransactionTest()
...
+    outputOperator2.beginWindow(1L);
+    outputOperator2.endWindow();
+
+    long windowIdOp = outputOperator.getStore().getCommittedWindowId(APP_ID, OPERATOR_ID);
+    Assert.assertEquals(0L, windowIdOp);
+
+    long windowIdOp2 = outputOperator2.getStore().getCommittedWindowId(APP_ID, OPERATOR_2_ID);
+    Assert.assertEquals(1L, windowIdOp2);
+
+    deleteOperator();
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: refactor deleteOperator to take operator as an argument and use it for outputOperator and outputOperstor2

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
address comments

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java (hunk 2 above)
Comment (oliverwnk) [reply library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java]: I agree

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java (hunk 1 above)
Comment (oliverwnk) [reply library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java]: Should have been '... required in the JMS provider'. 
What do you think about: "It is the responsibility of the user to create the meta-data queue in the JMS provider."

Comment (sanjaypujare) [reply library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java]: @oliverwnk your new wording looks good.

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 6]
```diff
@@ -46,11 +46,12 @@
   public static final String CLIENT_ID = "Client1";
   public static final String APP_ID = "appId";
   public static final int OPERATOR_ID = 1;
-  public static JMSBaseTransactionableStore store;
-  public static JMSStringSinglePortOutputOperator outputOperator;
+  public static final int OPERATOR_2_ID = 2;
+  //public static JMSStringSinglePortOutputOperator outputOperator;
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Any reason to keep the commented line around?

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 7]
```diff
@@ -106,94 +108,122 @@ private void createOperator()
...
-  /**
-   * This is a helper method to teardown a test operator.
-   */
-  private void deleteOperator()
-  {
-    outputOperator.teardown();
+    return outputOperator;
   }
 
   //@Ignore
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Can we delete this commented @Ignore line?

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 8]
```diff
@@ -106,94 +108,122 @@ private void createOperator()
...
+    JMSStringSinglePortOutputOperator jmsOutputOperator = createOperator(testOperatorContext);
+    JMSBaseTransactionableStore store = jmsOutputOperator.getStore();
+
     Assert.assertTrue("Should be connected.", store.isConnected());
-    deleteOperator();
+    jmsOutputOperator.teardown();
     Assert.assertFalse("Should not be connected.", store.isConnected());
   }
 
   //@Ignore
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Remove the commented line?

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 9]
```diff
@@ -106,94 +108,122 @@ private void createOperator()
...
     store.beginTransaction();
     Assert.assertTrue("Should be in transaction.", store.isInTransaction());
     store.commitTransaction();
     Assert.assertFalse("Should not be in transaction.", store.isInTransaction());
 
-    deleteOperator();
+    jmsOutputOperator.teardown();
   }
 
   //@Ignore
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Remove the line

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 10]
```diff
@@ -106,94 +108,122 @@ private void createOperator()
...
 
-    deleteOperator();
+    jmsOutputOperator.teardown();
+  }
+
+  /**
+   * Creates two operators with different operatorId and same appId to test correct functionality of storing and
+   * retrieving windowId with message selector
+   */
+  //@Ignore
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Remove

File: library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java [hunk 11]
```diff
@@ -106,94 +108,122 @@ private void createOperator()
...
+    Assert.assertEquals(0L, windowIdOp);
+
+    long windowIdOp2 = jmsOutputOperator2.getStore().getCommittedWindowId(APP_ID, OPERATOR_2_ID);
+    Assert.assertEquals(1L, windowIdOp2);
+
+    jmsOutputOperator.teardown();
+    jmsOutputOperator2.teardown();
   }
 
   ////@Ignore
```
Comment (sanjaypujare) [inline library/src/test/java/com/datatorrent/lib/io/jms/JMSTransactionableStoreTestBase.java]: Remove

**Review by sanjaypujare — CHANGES_REQUESTED** [review]
Most comments  are asking to remove commented code unless there is a reason to keep those lines (state what those reasons are)

**Comment by oliverwnk** [comment]
The '@Ignore' lines where not added by me, but I also see no reason to keep them in the code.

**Comment by sanjaypujare** [comment]
@PramodSSImmaneni  I have reviewed this but need someone to merge. Could you merge (after reviewing if needed)?

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java [hunk 12]
```diff
@@ -166,16 +175,15 @@ public void connect() throws IOException
...
-
       logger.debug("Base is null: {}", getBase() == null);
 
       if (getBase() != null) {
         logger.debug("Session is null: {}", getBase().getSession() == null);
       }
-
-      Queue queue = getBase().getSession().createQueue(queueName);
-      QueueBrowser browser = getBase().getSession().createBrowser(queue);
+      metaQueueName = getBase().getSubject() + ".metadata";
```
Comment (pramodin) [inline library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java:183]: Why not allow the user to also specify the name of this queue and default if not specified.

File: library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java [hunk 13]
```diff
@@ -186,16 +194,14 @@ public void connect() throws IOException
...
 
       connected = true;
       logger.debug("Connected. is in transaction: {}", inTransaction);
 
       if (!hasStore) {
         beginTransaction();
-        BytesMessage message = getBase().getSession().createBytesMessage();
-        message.writeLong(-1L);
-        producer.send(message);
+        writeWindowId(getAppId(), getOperatorId(), -1L);
```
Comment (pramodin) [inline library/src/main/java/com/datatorrent/lib/io/jms/JMSTransactionableStore.java:204]: Use Stateless.WINDOW_ID

**Comment by pramodin** [comment]
@sanjaypujare I have a couple of comments, please see. Also, squash and rebase. If @oliverwnk cannot work on this, please add "This closes #612" to the commit message.

**Comment by sanjaypujare** [comment]
@oliverwnk  let me know if you are going to be working on this.

**Pull request closed without merging**